src/GafferUI/AnimationGadget.cpp

```diff
@@ -434,55 +434,50 @@ void AnimationGadget::removeKeyframes()
 	requestRender();
 }
 
-// \todo: This seems a bit messy and it doesn't support Undo.
-// Not sure if we'd ideally hold pointers to the Keys instead to edit them in-place?
 void AnimationGadget::moveKeyframes( const V2f offset )
 {
+	// \todo: defaulting to no snapping, anticipating an extra mode for that at some point
+	float xSnappingOffset = offset.x; // offset relative to beginning of drag
+	if( m_snappingClosestKey.first.type != Animation::Invalid )
+	{
+		float unsnappedFrame = timeToFrame( m_snappingClosestKey.first.time + offset.x );
+		xSnappingOffset = frameToTime( round( unsnappedFrame ) ) - m_snappingClosestKey.first.time;
+	}
 
-	std::set<UniqueKey> tmp;
+	// \todo: can be retrieved once
+	KeyContainerIndex& index = m_selectedKeys.get<0>(); 
 
-	for( auto &uniqueKey : m_selectedKeys )
+	for( auto uniqueKeyIt = index.begin(); uniqueKeyIt != index.end(); ++uniqueKeyIt )
 	{
-		// create copy of Key from drag initiation
-		Animation::Key key = m_dragInitialKeys[uniqueKey];
 
-		// remove current key from the curve as it will need to change
-		uniqueKey.second->removeKey( uniqueKey.first.time );
+		uniqueKeyIt->second->removeKey( uniqueKeyIt->first.time );
 
-		// apply offsets
-		if( m_moveAxis == MoveAxis::Both || m_moveAxis == MoveAxis::Y )
+		// apply offset for key's value
+		if( m_moveAxis != MoveAxis::X )
 		{
-			key.value += offset.y;
+			index.modify( uniqueKeyIt, UniqueKeyChangeValue( uniqueKeyIt->first.value + offset.y ) );
```

Comment:
Note to self: This is currently adding the full offset every single time and needs to be fixed.